### PR TITLE
Persist: Conditional deletes + updates

### PR DESCRIPTION
**Pull request by snazy**

### CODE REVIEW

File: versioned/storage/bigtable/src/main/java/org/projectnessie/versioned/storage/bigtable/BigTablePersist.java [hunk 1]
```diff
@@ -540,6 +531,81 @@ public void upsertObjs(@Nonnull Obj[] objs) throws ObjTooLargeException {
...
+
+    Filter condition =
+        FILTERS
+            .chain()
+            .filter(
+                FILTERS
+                    .chain()
+                    .filter(FILTERS.qualifier().exactMatch(QUALIFIER_OBJ_VERS))
+                    .filter(FILTERS.value().exactMatch(obj.versionToken())))
+        // TODO adding this safety net doesn't work :(
```
Comment (adutra) [inline versioned/storage/bigtable/src/main/java/org/projectnessie/versioned/storage/bigtable/BigTablePersist.java:569]: Try this:

```java
ChainFilter objTypeFilter =
  FILTERS
      .chain()
      .filter(FILTERS.qualifier().exactMatch(QUALIFIER_OBJ_TYPE))
      .filter(FILTERS.value().exactMatch(OBJ_TYPE_VALUES.get(obj.type())));
ChainFilter objVersionFilter =
  FILTERS
    .chain()
    .filter(FILTERS.qualifier().exactMatch(QUALIFIER_OBJ_VERS))
    .filter(FILTERS.value().exactMatch(obj.versionToken()));
Filter condition =
    FILTERS.condition(objTypeFilter).then(objVersionFilter).otherwise(FILTERS.block());
```

File: versioned/storage/common-tests/src/main/java/org/projectnessie/versioned/storage/commontests/AbstractBasePersistTests.java [hunk 2]
```diff
@@ -1233,4 +1261,92 @@ public void nullHandlingInArrays() throws ObjTooLargeException, ObjNotFoundExcep
...
+    soft.assertThat(persist.deleteConditional(v2)).isFalse();
+    soft.assertThat(persist.fetchObj(v1.id())).isEqualTo(v1);
+    // exists, same version - delete must succeed
+    soft.assertThat(persist.deleteConditional(v1)).isTrue();
+    long t = System.currentTimeMillis();
+
+    soft.assertThatThrownBy(() -> persist.fetchObj(v1.id()))
+        .isInstanceOf(ObjNotFoundException.class);
+
+    if ("Cassandra".equals(persist.name())) {
```
Comment (adutra) [inline versioned/storage/common-tests/src/main/java/org/projectnessie/versioned/storage/commontests/AbstractBasePersistTests.java:1290]: Nit: maybe extract to a method, it's repeated twice.

File: versioned/storage/common-tests/src/main/java/org/projectnessie/versioned/storage/commontests/AbstractBasePersistTests.java [hunk 3]
```diff
@@ -1233,4 +1261,92 @@ public void nullHandlingInArrays() throws ObjTooLargeException, ObjNotFoundExcep
...
+    soft.assertThat(persist.updateConditional(v1, v2)).isTrue();
+    soft.assertThat(persist.fetchObj(v1.id())).isEqualTo(v2);
+
+    // exists, previous expected version - update must not succeed
+    soft.assertThat(persist.updateConditional(v1, v2)).isFalse();
+    soft.assertThat(persist.fetchObj(v1.id())).isEqualTo(v2);
+
+    // exists, expected version - update must succeed
+    soft.assertThat(persist.updateConditional(v2, v3)).isTrue();
+    soft.assertThat(persist.fetchObj(v1.id())).isEqualTo(v3);
```
Comment (adutra) [inline versioned/storage/common-tests/src/main/java/org/projectnessie/versioned/storage/commontests/AbstractBasePersistTests.java:1350]: Maybe test here the corner case where a versioned object exists, but with wrong type:

```java
    persist.deleteObj(v1.id());
    VersionedTestObj2 wrong = VersionedTestObj2.builder().from(v1).otherValue("blah").build();
    persist.storeObj(wrong);
    soft.assertThat(persist.updateConditional(v1, v2)).isFalse();
```

File: versioned/storage/common/src/main/java/org/projectnessie/versioned/storage/common/persist/UpdateableObj.java [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.projectnessie.versioned.storage.common.persist;
+
+import jakarta.annotation.Nonnull;
+
+public interface UpdateableObj extends Obj {
```
[the remark this answers is not in the document]
Comment (snazy) [reply versioned/storage/common/src/main/java/org/projectnessie/versioned/storage/common/persist/UpdateableObj.java:20]: Maybe not. It's not a "standard" type but a generic extension to `Obj`

File: versioned/storage/common/src/main/java/org/projectnessie/versioned/storage/common/json/ObjIdHelper.java [hunk 5]
```diff
@@ -31,13 +31,16 @@ public final class ObjIdHelper {
    */
   public static final String OBJ_ID_KEY = "nessie.storage.ObjId";
 
+  public static final String OBJ_VERS_KEY = "nessie.storage.ObjVersion";
+
   /**
    * Returns an {@link ObjectReader} for the given target {@link Obj} class and with the given
    * {@link ObjId} injectable under the key {@link #OBJ_ID_KEY}.
    */
   public static ObjectReader readerWithObjId(
```
[the remark this answers is not in the document]
Comment (snazy) [reply versioned/storage/common/src/main/java/org/projectnessie/versioned/storage/common/json/ObjIdHelper.java:40]: done

File: versioned/storage/common/src/main/java/org/projectnessie/versioned/storage/common/persist/UpdateableObj.java [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+  @Nonnull
+  String versionToken();
```
[the remark this answers is not in the document]
Comment (snazy) [reply versioned/storage/common/src/main/java/org/projectnessie/versioned/storage/common/persist/UpdateableObj.java:22]: Need to store it

[the remark this answers is not in the document]
Comment (snazy) [reply versioned/storage/common/src/main/java/org/projectnessie/versioned/storage/common/persist/UpdateableObj.java:22]: It's needed tho

File: versioned/storage/common-serialize/src/main/java/org/projectnessie/versioned/storage/serialize/SmileSerialization.java [hunk 7]
```diff
@@ -37,20 +37,27 @@ public final class SmileSerialization {
 
   private SmileSerialization() {}
 
-  public static Obj deserializeObj(ObjId id, byte[] data, String typeName, String compression) {
+  public static Obj deserializeObj(
```
[the remark this answers is not in the document]
Comment (snazy) [reply versioned/storage/common-serialize/src/main/java/org/projectnessie/versioned/storage/serialize/SmileSerialization.java:40]: done

File: versioned/storage/bigtable/src/main/java/org/projectnessie/versioned/storage/bigtable/BigTablePersist.java [hunk 8]
```diff
@@ -374,24 +376,13 @@ public boolean storeObj(@Nonnull Obj obj, boolean ignoreSoftSizeRestrictions)
                   Function.identity(), (ObjType type) -> ByteString.copyFromUtf8(type.name())));
 
   @NotNull
-  private ConditionalRowMutation mutationForStoreObj(
+  private ConditionalRowMutation conditionalForStoreObj(
```
[the remark this answers is not in the document]
Comment (snazy) [reply versioned/storage/bigtable/src/main/java/org/projectnessie/versioned/storage/bigtable/BigTablePersist.java:379]: done

File: versioned/storage/bigtable/src/main/java/org/projectnessie/versioned/storage/bigtable/BigTablePersist.java (hunk 1 above)
Comment (snazy) [reply versioned/storage/bigtable/src/main/java/org/projectnessie/versioned/storage/bigtable/BigTablePersist.java:569]: ah! this works! thanks!

File: versioned/storage/common-tests/src/main/java/org/projectnessie/versioned/storage/commontests/AbstractBasePersistTests.java (hunk 2 above)
Comment (snazy) [reply versioned/storage/common-tests/src/main/java/org/projectnessie/versioned/storage/commontests/AbstractBasePersistTests.java:1290]: done

File: versioned/storage/common-tests/src/main/java/org/projectnessie/versioned/storage/commontests/AbstractBasePersistTests.java (hunk 3 above)
Comment (snazy) [reply versioned/storage/common-tests/src/main/java/org/projectnessie/versioned/storage/commontests/AbstractBasePersistTests.java:1350]: Good idea, added

File: versioned/storage/common-proto/src/main/proto/org/projectnessie/versioned/storage/common/proto/storage.proto [hunk 9]
```diff
@@ -144,4 +144,5 @@ message CustomProto {
   string obj_type = 1;
   bytes data = 2;
   CompressionProto compression = 3;
+  optional string version_token = 4;
```
[the remark this answers is not in the document]
Comment (snazy) [reply versioned/storage/common-proto/src/main/proto/org/projectnessie/versioned/storage/common/proto/storage.proto:147]: (see above)

File: versioned/storage/common/src/main/java/org/projectnessie/versioned/storage/common/persist/UpdateableObj.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (snazy) [reply versioned/storage/common/src/main/java/org/projectnessie/versioned/storage/common/persist/UpdateableObj.java:22]: Agree, that's better.

File: versioned/storage/rocksdb/src/main/java/org/projectnessie/versioned/storage/rocksdb/RocksDBPersist.java [hunk 10]
```diff
@@ -442,6 +443,80 @@ public void upsertObjs(@Nonnull Obj[] objs) throws ObjTooLargeException {
...
+      TransactionDB db = b.db();
+      ColumnFamilyHandle cf = b.objs();
+      byte[] key = dbKey(id);
+
+      byte[] obj = db.get(cf, key);
+      if (obj == null) {
+        return false;
+      }
+      Obj existing = deserializeObj(id, obj);
+      if (!existing.type().equals(existing.type())) {
```
Comment (adutra) [inline versioned/storage/rocksdb/src/main/java/org/projectnessie/versioned/storage/rocksdb/RocksDBPersist.java:498]:
```suggestion
      if (!existing.type().equals(expected.type())) {
```

Comment (snazy) [reply versioned/storage/rocksdb/src/main/java/org/projectnessie/versioned/storage/rocksdb/RocksDBPersist.java]: meh - the old code had that check - just the `if` was dumb

File: versioned/storage/bigtable/src/main/java/org/projectnessie/versioned/storage/bigtable/BigTablePersist.java [hunk 11]
```diff
@@ -625,7 +686,10 @@ protected Obj computeNext() {
 
         List<RowCell> objCells = row.getCells(FAMILY_OBJS, QUALIFIER_OBJS);
         ByteBuffer obj = objCells.get(0).getValue().asReadOnlyByteBuffer();
-        return deserializeObj(id, obj);
+        List<RowCell> objVersionCells = row.getCells(FAMILY_OBJS, QUALIFIER_OBJ_VERS);
+        String versionToken =
```
Comment (adutra) [inline versioned/storage/bigtable/src/main/java/org/projectnessie/versioned/storage/bigtable/BigTablePersist.java:690]: Maybe extract to a method? Repeated 3 times.

File: versioned/storage/bigtable/src/main/java/org/projectnessie/versioned/storage/bigtable/BigTablePersist.java [hunk 12]
```diff
@@ -540,6 +524,83 @@ public void upsertObjs(@Nonnull Obj[] objs) throws ObjTooLargeException {
...
+    int indexSizeLimit =
+        ignoreSoftSizeRestrictions ? Integer.MAX_VALUE : effectiveIndexSegmentSizeLimit();
+    byte[] serialized = serializeObj(obj, incrementalIndexSizeLimit, indexSizeLimit, false);
+
+    Mutation mutation = Mutation.create();
+    objToMutation(obj, mutation::setCell, serialized);
+    return mutation;
+  }
+
+  static void objToMutation(Obj obj, ToMutation mutation, byte[] serialized) {
```
Comment (adutra) [inline versioned/storage/bigtable/src/main/java/org/projectnessie/versioned/storage/bigtable/BigTablePersist.java:586]: No need for a functional interface here, you can use `MutationApi<?> mutation`.

File: versioned/storage/rocksdb/src/main/java/org/projectnessie/versioned/storage/rocksdb/RocksDBPersist.java [hunk 13]
```diff
@@ -442,6 +444,86 @@ public void upsertObjs(@Nonnull Obj[] objs) throws ObjTooLargeException {
...
+      TransactionDB db = b.db();
+      ColumnFamilyHandle cf = b.objs();
+      byte[] key = dbKey(id);
+
+      byte[] bytes = db.get(cf, key);
+      if (bytes == null) {
+        return false;
+      }
+      Obj existing = deserializeObj(id, bytes, null);
+      if (!existing.type().equals(existing.type())) {
```
[the remark this answers is not in the document]
Comment (snazy) [reply versioned/storage/rocksdb/src/main/java/org/projectnessie/versioned/storage/rocksdb/RocksDBPersist.java:462]: TBH I don't understand your suggestion

File: versioned/storage/bigtable/src/main/java/org/projectnessie/versioned/storage/bigtable/BigTablePersist.java (hunk 12 above)
Comment (snazy) [reply versioned/storage/bigtable/src/main/java/org/projectnessie/versioned/storage/bigtable/BigTablePersist.java]: done + refactored

File: versioned/storage/bigtable/src/main/java/org/projectnessie/versioned/storage/bigtable/BigTablePersist.java (hunk 11 above)
Comment (snazy) [reply versioned/storage/bigtable/src/main/java/org/projectnessie/versioned/storage/bigtable/BigTablePersist.java]: done

File: versioned/storage/rocksdb/src/main/java/org/projectnessie/versioned/storage/rocksdb/RocksDBPersist.java (hunk 13 above)
[the remark this answers is not in the document]
Comment (adutra) [reply versioned/storage/rocksdb/src/main/java/org/projectnessie/versioned/storage/rocksdb/RocksDBPersist.java:462]: The if expression is always false.

[the remark this answers is not in the document]
Comment (snazy) [reply versioned/storage/rocksdb/src/main/java/org/projectnessie/versioned/storage/rocksdb/RocksDBPersist.java:462]: There's no such line in the code, no?

[the remark this answers is not in the document]
Comment (snazy) [reply versioned/storage/rocksdb/src/main/java/org/projectnessie/versioned/storage/rocksdb/RocksDBPersist.java]: argh!

**Review by adutra — APPROVED** [review]
